yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -331,14 +334,25 @@ private[spark] class ApplicationMaster(
                 finish(FinalApplicationStatus.FAILED,
                   ApplicationMaster.EXIT_REPORTER_FAILURE, "Exception was thrown " +
                     s"${failureCount} time(s) from Reporter thread.")
-
               } else {
                 logWarning(s"Reporter thread fails ${failureCount} time(s) in a row.", e)
               }
             }
           }
           try {
-            Thread.sleep(interval)
+            val numPendingAllocate = allocator.getNumPendingAllocate
+            if (numPendingAllocate > 0) {
+              currentAllocationInterval =
+                math.min(heartbeatInterval,currentAllocationInterval * 2)
+              logDebug(s"Number of pending allocations is ${numPendingAllocate}. " +
+                        "Sleeping for " + currentAllocationInterval)
+              Thread.sleep(currentAllocationInterval)
```

Comment:
@ehnalis if you're up for it, I'd be in favor of implementing something like Marcelo suggested in the current patch.  I think it would make this vastly more useful under dynamic allocation.